packages/dx-core/src/types.ts

```diff
@@ -0,0 +1,31 @@
+/*** makes all types in tuple readonly except functions
+ * MakeReadonly<[object, Function, string[]]> is equal to
+ * [ReadonlyObject<object>, Function, ReadonlyArray<string>]
+**/
+type ReadonlyTuple<T> = { readonly [K in keyof T]: Immutable<T[K]> };
+
+type Immutable<T> =
+// tslint:disable-next-line: array-type
+  T extends (infer P)[] ? ReadonlyArray<P> :
+  T extends Map<infer TKey, infer TValue> ? ReadonlyMap<TKey, TValue> :
+// tslint:disable-next-line: ban-types
+  T extends Function ? T :
+  T extends object ? ReadonlyObject<T> :
+  Readonly<T>;
+
+export type ReadonlyObject<T> = { readonly [K in keyof T]: Immutable<T[K]>; };
+
+type TupleHead<T> = T extends [infer U, ...any[]] ? U : never;
+
+export type PureReducer<S = any, P = any, R = S> = (
```

Comment:
What describe these letters? `S, P, R` If something important, maybe we should change them to more clear words?

Reply:
It's State, Payload, Result. Ok, I'll rename it